ludwig/config_validation/checks.py

```diff
@@ -0,0 +1,327 @@
+                raise ConfigValidationError(
+                    f"Trainer.optimizer.type must be one of: {list(optimizer_registry.keys())}."
+                )
+
+
+@register_config_check
+def check_feature_names_unique(config: ModelConfigDict) -> None:
+    """Checks that all feature names are unique."""
+    input_features = config[INPUT_FEATURES]
+    input_feature_names = {input_feature[NAME] for input_feature in input_features}
+
+    output_features = config[OUTPUT_FEATURES]
+    output_feature_names = {output_feature[NAME] for output_feature in output_features}
+
+    if len(input_feature_names) + len(output_feature_names) != len(input_features) + len(output_features):
+        raise ConfigValidationError("Feature names must be unique.")
+
+
+@register_config_check
+def check_tied_features_valid(config: ModelConfigDict) -> None:
+    """Checks that all tied features are valid."""
+    print("CHECKING TIED FEATURES!")
+    input_features = config[INPUT_FEATURES]
+    input_feature_names = {input_feature[NAME] for input_feature in input_features}
+
+    for input_feature in input_features:
+        if input_feature["tied"] and input_feature["tied"] not in input_feature_names:
+            raise ConfigValidationError(
+                f"Feature {input_feature[NAME]} is tied to feature {input_feature['tied']}, but the "
+                f"'{input_feature['tied']}' feature does not exist."
+            )
+
+
+@register_config_check
+def check_training_runway(config: ModelConfigDict) -> None:
+    """Checks that checkpoints_per_epoch and steps_per_checkpoint aren't simultaneously defined."""
+    if config[MODEL_TYPE] == MODEL_ECD:
+        if config[TRAINER]["checkpoints_per_epoch"] != 0 and config[TRAINER]["steps_per_checkpoint"] != 0:
+            raise ConfigValidationError(
+                "It is invalid to specify both trainer.checkpoints_per_epoch AND "
+                "trainer.steps_per_checkpoint. Please specify one or the other, or specify neither to "
+                "checkpoint/eval the model every epoch."
+            )
+
+
+@register_config_check
+def check_gbm_horovod_incompatibility(config: ModelConfigDict) -> None:
+    """Checks that GBM model type isn't being used with the horovod backend."""
+    if BACKEND not in config:
+        return
+    if config[BACKEND] is None:
+        return
+    if config[MODEL_TYPE] == MODEL_GBM and config[BACKEND][TYPE] == "horovod":
+        raise ConfigValidationError("Horovod backend does not support GBM models.")
+
+
+@register_config_check
+def check_gbm_single_output_feature(config: ModelConfigDict) -> None:
+    """GBM models only support a single output feature."""
+    if config[MODEL_TYPE] == MODEL_GBM:
+        if len(config[OUTPUT_FEATURES]) != 1:
+            raise ConfigValidationError("GBM models only support a single output feature.")
+
+
+@register_config_check
+def check_gbm_feature_types(config: ModelConfigDict) -> None:
```

Comment:
This is also in the schema already. Remove.

Reply:
+1